### DISCUSSION TITLE
Provide Desktop Mode

### DISCUSSION THREAD

**Discussion by rodja** [zauberzeug/nicegui#89] [Ideas / Feature Requests]
I just stumbled on https://github.com/Widdershin/flask-desktop/ which provides a desktop wrapper around a flask application. It uses Qt (through https://pypi.org/project/PySide2/) to open a window and display the WebView. 

We could do something similar in NiceGUI. Maybe activation is as simple as an additional parameter `desktop=True` in the `ui.run` call?

**Comment by rodja** [comment]
Maybe an implementation with https://pypi.org/project/pywebview/ would be even nicer...

**Reply by eli-kha** [reply]
I made a POC for pywebview with NiceGUI. It's a bit long so I made a gist.
https://gist.github.com/eli-kha/06a47bfdf1e50f4cdfc3f43a199a6d2d

I couldn't get it to work with `ui.run` but I didn't try much. I suspect the issue was related to the reload functionality.

I tend to think it's not something that you can or should embed into `ui.run`.
A very simple "just show this URL" can be implemented but if you want to move between urls or have multiple windows it can get very complicated very fast.

Maybe just an example reference code?

**Reply by rodja** [reply]
Thanks for bringing this up again. I think I found the shortest possible demo yet:

```py
from nicegui import ui
import webview # pip install pywebview

ui.button('button', on_click=lambda: ui.notify('Nice!'))

window = webview.create_window('NiceGUI', url='http://localhost:8080')
webview.start(lambda _: ui.run(show=False, reload=False), window)
```

**Reply by rodja** [reply]
Ah, it may be a bit longer if you also want to close the window:

```py
from nicegui import ui
import webview # pip install pywebview

ui.button('button', on_click=lambda: ui.notify('Nice!'))

window = webview.create_window('NiceGUI', url='http://localhost:8080')
window.events.closing += app.shutdown
webview.start(lambda _: ui.run(show=False, reload=False), window)
```

**Reply by eli-kha** [reply]
Much better than mine!
Your code didn't work for me due to 2 issues: 
- missing import of app
- permissions problem for the webview storage (cookies and such)

Here's a slightly revised example:
```python
import tempfile

from nicegui import ui, app
import webview # pip install pywebview

ui.button('button', on_click=lambda: ui.notify('Nice!'))

window = webview.create_window('NiceGUI', url='http://localhost:8080')
window.events.closing += app.shutdown
webview.start(lambda _: ui.run(show=False, reload=False), window, storage_path=tempfile.mkdtemp())
```

**Comment by rodja** [comment]
Maybe this will be implemented as part of #246

**Reply by rodja** [reply]
No, it wont.

**Comment by eezis** [comment]
A Progressive Web App (PWA) would be brilliant, and that is the first thing I thought of when I saw the examples. If NiceGUI is leveraging Quasar -- which is a great platform -- then it would seem like there's a built in pathway to producing PWAs?

**Reply by rodja** [reply]
I'm not so sure (see https://github.com/zauberzeug/nicegui/issues/246#issuecomment-1386451204). But @schmurian will look into this. Maybe I'm missing something right now. Happy to learn.

**Comment by eli-kha** [comment]
Hi!

I wanted to share that I found https://github.com/ClimenteA/flaskwebgui, which builds on top of flask-desktop.
To do a quick test I replaced the run in the FastAPI example and it run without any problems:

```python
if __name__ == '__main__':
    from flaskwebgui import FlaskUI

    # uvicorn.run(app, host='0.0.0.0', port=8000)
    FlaskUI(app=app, server="fastapi").run()
```

I didn't test it beyond what I just showed but I'll be experimenting with it at work and I'll update if I have insights.

**Reply by al-eax** [reply]
that's pretty cool and works well:

![grafik](https://user-images.githubusercontent.com/2324292/220322937-9f655ec0-2d2f-428a-85ed-f863870122db.png)

```py
from fastapi import FastAPI
from flaskwebgui import FlaskUI
from nicegui import ui

app = FastAPI()

ui.label("Hello World")
ui.button("click me", on_click=lambda: ui.notify("clicked button"))

ui.run_with(app,dark=True)

if __name__ == "__main__":
    FlaskUI(app=app, server="fastapi").run()
```

**Reply by AL-THE-BOT-FATHER** [reply]
This is pretty awesome. 
How could you bounce this down to an executable? 
pyinstaller?

**Reply by al-eax** [reply]
im guess you can,
flaskwebguis github says:

> You can distribute it as a standalone desktop app with pyinstaller or [pyvan](https://github.com/ClimenteA/pyvan). If pyinstaller failes try pyinstaller version 5.6.2.

and nicegui can be packed into pyinstaller as well:
https://nicegui.io/reference#package_for_installation

**Reply by smojef** [reply]
This is great!  Unfortunately, when trying the code above on Windows, the frame opened full screen and I got this popup:
![flaskwebgui](https://user-images.githubusercontent.com/43252951/221103246-bb9a5f48-8a01-4582-befe-7ff3d5a90fbe.jpg)
However, after declining, it worked well, even after restart.

**Reply by al-eax** [reply]
Maybe the sync message can be turned off by adding `--disable-sync` command line argument for chrome based browsert.

**Reply by bobwatcherx** [reply]
> that's pretty cool and works well:
> 
> ![grafik](https://user-images.githubusercontent.com/2324292/220322937-9f655ec0-2d2f-428a-85ed-f863870122db.png)
> 
> ```python
> from fastapi import FastAPI
> from flaskwebgui import FlaskUI
> from nicegui import ui
> 
> app = FastAPI()
> 
> ui.label("Hello World")
> ui.button("click me", on_click=lambda: ui.notify("clicked button"))
> 
> ui.run_with(app,dark=True)
> 
> if __name__ == "__main__":
>     FlaskUI(app=app, server="fastapi").run()
> ```

how to be able to hot reload when saved in desktop mode

**Reply by falkoschindler** [reply]
I _think_ hot reload is not possible.

**Comment by morningstarsabrina** [comment]
it will be nice to have a desktop option :)
who know maybe one day we will have android apk option just like flutter ;)

**Comment by gloomie** [comment]
+1 for desktop mode. A way to embed this in flutter (like eg Flet does or as a webview) would be also very interesting. Otherwise a PWA solution would be exciting too...

**Comment by hditano** [comment]
Please!!. I really want to get something else than Eel!

**Reply by rodja** [reply]
Have you tried the solution in https://github.com/zauberzeug/nicegui/discussions/89#discussioncomment-5062741? If it works for multiple people, it could be added to the documentation.

**Reply by hditano** [reply]
it does work... but I found something that I cannot replace.

- Hide titlebar ( window) - eel: no_titlebar = True

![image](https://user-images.githubusercontent.com/6069906/223362149-b6b1c7b8-e56f-4d26-b944-e935263a9b6d.png)


- able to drag around the window setting an ID to an html element

![Animation](https://user-images.githubusercontent.com/6069906/223364060-6aecba4b-0d13-4b36-adb9-5608a3010b1e.gif)

- server wont start if you want to hide console through pyinstaller parameters: --windowed or --noconsole.

**Reply by ItsCubeTime** [reply]
Im also looking for a non-titlebar solution. In my case I need to keep the window resize controls as well.

**Reply by eli-kha** [reply]
see my latest comment: https://github.com/zauberzeug/nicegui/discussions/89#discussioncomment-5277898

**Comment by ItsCubeTime** [comment]
Another project worth considering is Toga. Which also provides mobile support with webview for all target platforms https://beeware.org/project/projects/libraries/toga/

Doesnt support topbar-less windows for desktop however & it seems the developers have little interest implementing it (in fact, they used closed a PR where I implemented this particular feature) https://github.com/beeware/toga/issues/733 Other than that it seems like a nice, lean project

**Comment by bobwatcherx** [comment]
hi this code is working for desktop app .  but there is only one problem, namely .  every change and reload .  it will open a new window.  maybe it can be improved a little.  to work.  if this works .  I like it better than in the browser
```py
from flaskwebgui import FlaskUI
from nicegui import ui

ui.label("Hello Super NiceGUI  ssss!")
ui.button("ccknewqnl", on_click=lambda: ui.notify("button was pressed"))

def start_nicegui(**kwargs):
    ui.run(**kwargs)

if __name__ in {"__main__", "__mp_main__"}:
    DEBUG = False

    if DEBUG:
        ui.run()
    else:
        FlaskUI(
            server=start_nicegui,
            server_kwargs={"reload": False, "show": False, "port": 3000},
            width=800,
            height=600,
        ).run()
```

**using command flet main.py**

**Comment by rodja** [comment]
@eli-kha and me managed to build a full working example including auto-reload and graceful shutdown with [pywebview](https://pywebview.flowrl.com/):

```py
import multiprocessing
import os
import signal
import tempfile
import webview  # pip install webview
from nicegui import ui, app

ui.button('button', on_click=lambda: ui.notify('Nice!'))

def open_window(event):
    window = webview.create_window('NiceGUI', url='http://localhost:8080')
    window.events.closing += event.set  # signal that the program should be closed to the main process
    webview.start(storage_path=tempfile.mkdtemp())


# create shutdown event for inter process communication
shutdown = multiprocessing.Event()
# start webview in a subprocess
app.on_startup(lambda: multiprocessing.Process(target=open_window, args=(shutdown,), daemon=True).start())
#  repeatedly check if the program should be terminated
ui.timer(0.1, lambda: os.kill(os.getpgid(os.getpid()), signal.SIGTERM) if shutdown.is_set() else None)

ui.run(show=False)
```

**Comment by rodja** [comment]
With a small modification, the application window does not close/open for every reload:

```py
#!/usr/bin/env python3
import multiprocessing
import os
import signal
import tempfile
import webview  # pip install webview
from nicegui import ui

ui.button('button', on_click=lambda: ui.notify('Nice!'))


def open_window(event):
    window = webview.create_window('NiceGUI', url='http://localhost:8080')
    window.events.closing += event.set  # signal that the program should be closed to the main process
    webview.start(storage_path=tempfile.mkdtemp())

# create shutdown event for inter process communication
shutdown = multiprocessing.Event()
#  repeatedly check if the program should be terminated
ui.timer(0.1, lambda: os.kill(os.getpgid(os.getpid()), signal.SIGTERM) if shutdown.is_set() else None)
# start webview in a subprocess once when the program starts (not when uvicorn parses the file again with __name__ == '__mp_main__')
if __name__ == "__main__":
    multiprocessing.Process(target=open_window, args=(shutdown,), daemon=False).start()

ui.run(show=False)
```

If this works on all platforms I would really like to have this integrated into core NiceGUI with `ui.run(desktop=true)` or similar.

**Reply by bobwatcherx** [reply]
i got an error from your code.  i use linux
```
root@minx-Lenovo-V145-14AST:/home/minx/belajar/nicePro/appdekstop# python3 test.py 
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/minx/belajar/nicePro/appdekstop/test.py", line 13, in open_window
    window = webview.create_window('NiceGUI', url='http://localhost:8080')
AttributeError: module 'webview' has no attribute 'create_window'
NiceGUI ready to go on http://0.0.0.0:8080
```

**Reply by phoskee** [reply]
Try `pip install pywebview`

**Reply by bobwatcherx** [reply]
it works fine on linux .  but is there a shorter and simpler way than that long code. @rodja

**Reply by rodja** [reply]
You mean with FlaskWebUI as you posted in https://github.com/zauberzeug/nicegui/discussions/89#discussioncomment-5279588?
I've had a short look in the code of the project. They are not starting a real Windows with a webview but rather a browser window. I think we have more opportunities (like adding a native menu etc.) if we use a lib which wraps the platform specific apis.

**Comment by rodja** [comment]
I just created PR #519 which introduces the feature as "standalone" and "fullscreen" flags to `ui.run`.

**Comment by rodja** [comment]
The PR #519 is merged. But I'm still a bit unhappy with the run parameter "standalone". I chose it above "desktop_mode" because maybe we will get Android/iOS support later on (fingers crossed). Because 1.2.0 is not jet released we could still change the name without breaking the API. What do you suggest? I'm currently leaning towards `ui.run(native_window=True)`.

**Reply by falkoschindler** [reply]
What about a `Literal` type instead of a boolean, like

```py
mode: Literal['browser', 'native'] = 'browser'
```

or

```py
output: Literal['browser', 'native'] = 'browser'
```

This would allow new modes like "mobile" in the future.

**Reply by rodja** [reply]
I don't think there needs to be a separate mode for "mobile". I will still be "native", but running on a different device. Also we currently also allow configuring the size of the view by passing a tuple to the parameter. Btw: my new favorite name for the parameter is `native_view`.

**Reply by falkoschindler** [reply]
In https://github.com/zauberzeug/nicegui/commit/cf92ad18f2f1e84eefee459c1f30ffea2f9802e7 we changed the API as follows:

```py
native: bool = False,
window_size: Optional[Tuple[int, int]] = None,
fullscreen: bool = False,
```

**Comment by rodja** [comment]
We finally published v1.2.0 which paves the way for NiceGUI to be the Electron of Python: https://github.com/zauberzeug/nicegui/releases/tag/v1.2.0

**Comment by bobwatcherx** [comment]
for desktop mode how to do custom bar . like frameless windows. so for the bar we can modify it. including I will make my own minimize maximize button and close button Like QBar in quasars  And if possible can you add it to the nicegui documentation if possible thanks

**Reply by rodja** [reply]
Try the new `app.native.windows_args` and `app.native.start_args` dictionaries. They are described in https://nicegui.io/documentation#native_mode.

**Reply by wanghaisheng** [reply]
![image](https://user-images.githubusercontent.com/2363295/235623739-22a399e7-2daa-4e3f-ab8e-b41890c1a1a0.png)
@rodja   a little modification to demo code, failed

**Reply by wanghaisheng** [reply]
with a fresh new py, 'app running in native mode' not show either![image](https://user-images.githubusercontent.com/2363295/235625807-ceff97b5-28e0-48ae-883a-2a3072bdaffd.png)

**Reply by falkoschindler** [reply]
Strange. This code works on my machine:
```py
from nicegui import app, ui

ui.label('app running in native mode')
app.native.window_args['resizable'] = False
app.native.start_args['debug'] = True

ui.run(native=True, window_size=(400, 300), fullscreen=False)
```

**Reply by phoskee** [reply]
Same here, it works
Probably you have updated some non compatible dependent packages, same thing happened to me

**Reply by phoskee** [reply]
@bobwatcherx take a look to #770

**Reply by wanghaisheng** [reply]
do you need more debug info I  can provide

**Reply by phoskee** [reply]
In your screenshots you can see that you did a package update, or installed something, how?
Try to share the analysis of the web page (right click, inspect or analyze)

**Reply by wanghaisheng** [reply]
I  have a new fresh conda env ,
the same issue with inspect  
![image](https://user-images.githubusercontent.com/2363295/235649152-cba4c737-a2af-46a1-9d2d-4036c5cfdf1d.png)
